agent-instrumentation/src/jvmMain/kotlin/com/epam/drill/agent/instrument/reactor/ProxyUtils.kt

```diff
@@ -0,0 +1,100 @@
+import net.bytebuddy.TypeCache
+import net.bytebuddy.description.modifier.Visibility
+import net.bytebuddy.dynamic.DynamicType
+import net.bytebuddy.dynamic.loading.ClassLoadingStrategy
+import net.bytebuddy.implementation.MethodDelegation
+import net.bytebuddy.implementation.bind.annotation.*
+import net.bytebuddy.matcher.ElementMatchers.*
+import java.util.function.Function
+
+const val DRILL_DELEGATE_FIELD = "drillDelegate"
+const val DRILL_REQUEST_FIELD = "drillRequest"
+const val DRILL_CONTEXT_FIELD = "drillContext"
+const val DRILL_CONTEXT_KEY = "drillRequest"
+const val SUBSCRIPTION_CLASS = "org.reactivestreams.Subscription"
+const val SUBSCRIBER_CLASS = "reactor.core.CoreSubscriber"
+
+/**
+ * The cache of proxy classes
+ */
+val proxyClassCache = TypeCache<Class<*>>()
+
+/**
+ * Creates a proxy class and an instance of this class to intercept calls with this interceptor.
+ * Every public method of the class will be intercepted by the interceptor.
+ * @param delegate the original instance of the class.
+ * @param clazz the class that will be delegated. Must be a superclass of the class of the delegate.
+ * @param interceptor the Byte buddy method interceptor.
+ * @param configure the Byte buddy configuration which will be applied before building proxy class.
+ * @param initialize the proxy instance initialization logic.
+ * @return the proxy instance.
+ */
+@Suppress("UNCHECKED_CAST")
+inline fun <T> createProxyDelegate(
+    delegate: Any,
+    clazz: Class<T>,
+    interceptor: Any,
+    crossinline configure: DynamicType.Builder.FieldDefinition.Optional.Valuable<*>.() -> DynamicType.Builder.FieldDefinition.Optional.Valuable<*> = { this },
+    crossinline initialize: (T, Class<T>) -> Unit = { _, _ -> }
+): T {
+    val proxyType = proxyClassCache.findOrInsert(clazz.classLoader, clazz) {
+        ByteBuddy()
+            .subclass(clazz)
+            .defineField(DRILL_DELEGATE_FIELD, clazz, Visibility.PUBLIC)
+            .let(configure)
+            .method(isPublic())
+            .intercept(MethodDelegation.withDefaultConfiguration()
+                .withBinders(Pipe.Binder.install(Function::class.java))
+                .to(interceptor))
+            .make()
+            .load(clazz.classLoader, ClassLoadingStrategy.Default.INJECTION)
+            .loaded
+    } as Class<T>
+    val proxy = proxyType.getConstructor().newInstance()
+    proxyType.getField(DRILL_DELEGATE_FIELD).set(proxy, delegate)
+    initialize(proxy, proxyType)
+    return proxy
+}
+
+/**
+ * Propagates the drill request to the given body via the ThreadLocal context.
+ * If the request was already propagated, the previous request will be restored after the body invocation.
+ * @param ctx the drill request
+ * @param body the body function in which the drill request will be propagated
+ * @return the result of the body function
+ */
+inline fun <T> propagateDrillRequest(ctx: DrillRequest, requestHolder: RequestHolder, body: () -> T?): T? {
```

Comment:
1. Can we rename all instances of `"drillRequest"` and  to `"drillContext"`? Request sounds very much bound to HTTP, and to _request_. Meanwhile, in the future, we might want to:
- write it to "response" headers
- write to a different transport altogether

Both cases will be very confusing.

Maybe do that _after_ merge

2. `to the given body` - `to the given "body" lambda expression`
3. To be precise, it's propagated to the thread in which lambda is executed, is that right?

Reply:
+ rename "RequestHolder"